blockchain/stake/internal/tickettreap/common.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2015-2016 The btcsuite developers
+// Copyright (c) 2016 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package tickettreap
+
+import (
+	"math/rand"
+	"time"
+
+	"github.com/decred/dcrd/chaincfg/chainhash"
+)
+
+const (
+	// staticDepth is the size of the static array to use for keeping track
+	// of the parent stack during treap iteration.  Since a treap has a very
+	// high probability that the tree height is logarithmic, it is
+	// exceedingly unlikely that the parent stack will ever exceed this size
+	// even for extremely large numbers of items.
+	staticDepth = 128
+
+	// nodeFieldsSize is the size the fields of each node takes excluding
+	// the contents of the key and value.  It assumes 64-bit pointers so
+	// technically it is smaller on 32-bit platforms, but overestimating the
+	// size in that case is acceptable since it avoids the need to import
+	// unsafe.  It consists of 8 bytes for each of the value, priority,
+	// left, and right fields (8*4).
+	nodeFieldsSize = 32
+
+	// nodeValueSize is the size of the fixed-size fields of a Value.
+	nodeValueSize = 4
+)
+
+var (
+	// rng is a new random number generator that is local to the package.
+	rng = rand.New(rand.NewSource(time.Now().UnixNano()))
```

Comment:
rand.Rand's methods aren't safe for concurrent access (unlike the global RNG) so this needs a mutex (either on the variable itself, or lock the source like lockedSource in the stdlib).